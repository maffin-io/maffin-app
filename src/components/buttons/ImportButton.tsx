import React from 'react';
import Modal from 'react-modal';
import { BiImport } from 'react-icons/bi';

import DBImportButton from './import/DBImportButton';
import PlaidImportButton from './import/PlaidImportButton';

export interface ImportButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  className?: string;
  onImport?: Function;
}

export default function ImportButton({
  className = 'btn btn-primary',
  onImport,
  ...props
}: ImportButtonProps): JSX.Element {
  const [isModalOpen, setIsModalOpen] = React.useState<boolean>(false);

  return (
    <>
      {/* @ts-ignore */}
      <Modal
        isOpen={isModalOpen}
        overlayClassName="overlay"
        className="modal bg-background-800"
      >
        <button
          type="button"
          className="float-right"
          onClick={() => setIsModalOpen(false)}
        >
          X
        </button>
        <div className="grid grid-cols-2 justify-center mt-5">
          <DBImportButton onImport={() => setIsModalOpen(false)} />
          <PlaidImportButton onImport={() => setIsModalOpen(false)} />
        </div>
      </Modal>
      <button
        id="import-button"
        type="button"
        className={className}
        onClick={() => setIsModalOpen(!isModalOpen)}
        {...props}
      >
        <BiImport className="inline-block align-middle mr-1" />
        <span className="inline-block align-middle">Import</span>
      </button>
    </>
  );
}
