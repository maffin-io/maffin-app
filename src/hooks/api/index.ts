import { DateTime } from 'luxon';
import { mutate, SWRResponse } from 'swr';
import useSWRImmutable from 'swr/immutable';

import { Commodity } from '@/book/entities';
import { InvestmentAccount } from '@/book/models';
import * as queries from '@/lib/queries';
import type { PriceDBMap } from '@/book/prices';
import type { AccountsMap } from '@/types/book';
import type { Account, Split, Transaction } from '@/book/entities';

export function useStartDate(): SWRResponse<DateTime> {
  const key = '/api/start-date';
  return useSWRImmutable(
    key,
    fetcher(queries.getEarliestDate, key),
  );
}

export function useMainCurrency(): SWRResponse<Commodity> {
  const key = '/api/main-currency';
  return useSWRImmutable(
    key,
    fetcher(queries.getMainCurrency, key),
  );
}

export function useCommodity(guid: string): SWRResponse<Commodity> {
  const key = `/api/commodities/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => Commodity.findOneByOrFail({ guid }), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useCommodities(): SWRResponse<Commodity[]> {
  const key = '/api/commodities';

  const result = useSWRImmutable(
    key,
    fetcher(() => Commodity.find(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    result.data.forEach(
      (commodity: Commodity) => mutate(
        `/api/commodities/${commodity.guid}`,
        commodity,
        { revalidate: false },
      ),
    );
  }

  return result;
}

export function useLatestTxs(): SWRResponse<Transaction[]> {
  const key = '/api/txs/latest';
  return useSWRImmutable(
    key,
    fetcher(queries.getLatestTxs, key),
  );
}

export function useAccount(guid: string): SWRResponse<Account> {
  const key = `/api/accounts/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getAccounts(guid), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useAccounts(): SWRResponse<AccountsMap> {
  const key = '/api/accounts';

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getAccounts(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    Object.values(result.data as AccountsMap).forEach(
      (account: Account) => mutate(`/api/accounts/${account.guid}`, account, { revalidate: false }),
    );
  }

  return result;
}

export function useAccountsMonthlyTotals(): SWRResponse<queries.MonthlyTotals> {
  const { data: accounts } = useAccounts();
  const { data: prices } = useSWRImmutable(
    '/api/prices',
    fetcher(() => queries.getPrices({}), '/api/prices'),
  );

  const key = '/api/monthly-totals';
  const result = useSWRImmutable(
    (accounts && prices) ? key : null,
    fetcher(
      () => queries.getMonthlyTotals(accounts as AccountsMap, prices),
      key,
    ),
  );

  return result;
}

export function useInvestment(guid: string): SWRResponse<InvestmentAccount> {
  const key = `/api/investments/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getInvestment(guid), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useInvestments(): SWRResponse<InvestmentAccount[]> {
  const key = '/api/investments';

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getInvestments(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    result.data.forEach(
      (investment: InvestmentAccount) => mutate(
        `/api/investments/${investment.account.guid}`,
        investment,
        { revalidate: false },
      ),
    );
  }

  return result;
}

export function useSplits(guid: string): SWRResponse<Split[]> {
  const key = `/api/splits/${guid}`;
  return useSWRImmutable(
    key,
    fetcher(async () => queries.getSplits(guid), key),
  );
}

/**
 * Returns prices for a given commodity
 */
export function usePrices(c: Commodity | undefined): SWRResponse<PriceDBMap> {
  const key = `/api/prices/${c?.guid}`;
  return useSWRImmutable(
    c?.guid ? key : null,
    fetcher(async () => queries.getPrices({ from: c }), key),
  );
}

function fetcher(f: () => Promise<any>, key: string) {
  return async () => {
    const start = performance.now();
    const r = await f();
    const end = performance.now();
    console.log(`${key}: ${end - start}ms`);
    return r;
  };
}
