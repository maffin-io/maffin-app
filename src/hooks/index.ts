export { default as useDataSource, DataSourceContext } from './useDataSource';
export type { DataSourceContextType } from './useDataSource';
export { default as useBookStorage } from './useBookStorage';
export { default as useGapiClient } from './useGapiClient';
