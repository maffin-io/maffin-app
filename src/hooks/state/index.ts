export { useTheme } from './useTheme';
export { useOnline } from './useOnline';
