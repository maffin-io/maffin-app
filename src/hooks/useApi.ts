import useSWRImmutable from 'swr/immutable';
import { Fetcher, SWRResponse, SWRConfiguration } from 'swr';

import { Commodity } from '@/book/entities';
import * as queries from '@/lib/queries';
import { PriceDB } from '@/book/prices';
import getUser from '@/lib/getUser';

/**
 * Wrapper around SWR to make the usable keys explicit and re-usable
 */
export default function useApi(key: ApiPaths | null): SWRResponse {
  const f = key !== null ? API[key].f : () => {};
  const swrArgs = key !== null ? API[key].swrArgs : undefined;

  return useSWRImmutable(key, f, swrArgs);
}

export type ApiPaths = (
  '/api/start-date'
  | '/api/main-currency'
  | '/api/commodities'
  | '/api/accounts'
  | '/api/investments'
  | '/api/prices/today'
  | '/api/txs/latest'
  | '/api/user'
);

const API: {
  [key in ApiPaths]: {
    f: Fetcher,
    swrArgs?: SWRConfiguration,
  };
} = {
  '/api/start-date': {
    f: queries.getEarliestDate,
  },
  '/api/main-currency': {
    f: queries.getMainCurrency,
  },
  '/api/commodities': {
    f: Commodity.find,
  },
  '/api/accounts': {
    f: async (k: string) => {
      const start = performance.now();
      const r = await queries.getAccounts();
      const end = performance.now();
      console.log(`${k}: ${end - start}ms`);
      return r;
    },
  },
  '/api/investments': {
    f: queries.getInvestments,
  },
  '/api/prices/today': {
    f: PriceDB.getTodayQuotes,
  },
  '/api/txs/latest': {
    f: queries.getLatestTxs,
  },
  '/api/user': {
    f: getUser,
    swrArgs: {
      refreshInterval: 100000,
      revalidateOnMount: true,
    },
  },
};
