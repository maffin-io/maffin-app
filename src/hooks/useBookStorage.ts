import React from 'react';

import useGapiClient from '@/hooks/useGapiClient';
import type BookStorage from '@/lib/storage/BookStorage';
import {
  GDriveBookStorage,
  DemoBookStorage,
} from '@/lib/storage';
import useSession from './useSession';

type UseBookStorageReturn = {
  storage: BookStorage | null,
};

export default function useBookStorage(): UseBookStorageReturn {
  const [isGapiLoaded] = useGapiClient();
  const [storage, setStorage] = React.useState<BookStorage | null>(null);
  const { roles } = useSession();

  const [error, setError] = React.useState<Error | null>(null);
  if (error) {
    throw error;
  }

  React.useEffect(() => {
    async function load() {
      let instance: BookStorage = new DemoBookStorage();

      if (roles.isPremium) {
        instance = new GDriveBookStorage(window.gapi.client);
      }

      try {
        await instance.initStorage();
        setStorage(instance);
      } catch (e) {
        setError(e as Error);
      }
    }

    if (isGapiLoaded) {
      load();
    }
  }, [isGapiLoaded, roles.isPremium]);

  return { storage };
}
