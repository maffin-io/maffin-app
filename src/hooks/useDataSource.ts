import React from 'react';
import { DataSource } from 'typeorm';
import initSqlJs from 'sql.js';
import { mutate } from 'swr';
import pako from 'pako';

import { insertTodayPrices } from '@/lib/Stocker';
import useBookStorage from '@/hooks/useBookStorage';
import { importBook as importGnucashBook } from '@/lib/gnucash';
import {
  Account,
  Book,
  Commodity,
  Price,
  Split,
  Transaction,
} from '@/book/entities';
import { MIGRATIONS } from '@/book/migrations';
import { isStaging } from '@/helpers/env';
import { useQueryClient } from '@tanstack/react-query';
import type BookStorage from '@/lib/storage/BookStorage';

export type DataSourceContextType = {
  datasource: DataSource | null,
  save: () => Promise<void>,
  importBook: (rawBook: Uint8Array) => Promise<void>,
  isLoaded: boolean,
};

const DATASOURCE: DataSource = new DataSource({
  type: 'sqljs',
  synchronize: true,
  logging: false,
  entities: [Account, Book, Commodity, Price, Split, Transaction],
  migrations: MIGRATIONS,
});

export const DataSourceContext = React.createContext<DataSourceContextType>({
  datasource: DATASOURCE,
  save: async () => {},
  importBook: async () => {},
  isLoaded: false,
});

export default function useDataSource(): DataSourceContextType {
  const { storage } = useBookStorage();
  const [isLoaded, setIsLoaded] = React.useState(DATASOURCE.isInitialized);
  const queryClient = useQueryClient();

  React.useEffect(() => {
    async function load() {
      let rawBook: Uint8Array;
      if (storage && !DATASOURCE.isInitialized && queryClient) {
        ([, rawBook] = await Promise.all([
          initSqlJs({ locateFile: file => `/${file}` }),
          storage.get(),
        ]));

        // Due to async nature, this value can change between this and the previous statement
        // as it's a shared global
        if (!DATASOURCE.isInitialized) {
          DATASOURCE.setOptions({
            type: 'sqljs',
            synchronize: true,
            logging: false,
            entities: [Account, Book, Commodity, Price, Split, Transaction],
            migrations: MIGRATIONS,
            extra: {
              queryClient,
            },
          });
          const start = performance.now();
          await DATASOURCE.initialize();

          if (rawBook.length) {
            await DATASOURCE.sqljsManager.loadDatabase(rawBook);
            await DATASOURCE.runMigrations();
          } else {
            await createEmptyBook();
          }
          const end = performance.now();
          console.log(`init datasource: ${end - start}ms`);
        }
        await loadStockerPrices();
        setIsLoaded(true);
      }
    }

    load();
  }, [storage, queryClient]);

  if (!isLoaded) {
    return {
      datasource: DATASOURCE,
      save: async () => {},
      importBook: async () => {},
      isLoaded: false,
    };
  }

  return {
    datasource: DATASOURCE,
    save: () => save(storage as BookStorage),
    importBook: (rawData: Uint8Array) => importBook(storage as BookStorage, rawData),
    isLoaded,
  };
}

async function save(storage: BookStorage) {
  mutate('/state/save', true, { revalidate: false });
  const rawBook = DATASOURCE.sqljsManager.exportDatabase();
  await storage.save(rawBook);
  mutate('/state/save', false, { revalidate: false });
}

async function importBook(storage: BookStorage, rawData: Uint8Array) {
  let parsedData;
  try {
    parsedData = pako.ungzip(rawData);
  } catch (err) {
    parsedData = rawData;
  }

  const rawBook = await importGnucashBook(parsedData);
  await DATASOURCE.sqljsManager.loadDatabase(rawBook);

  DATASOURCE.options.extra.queryClient.refetchQueries({
    type: 'all',
  });

  await loadStockerPrices();
  await save(storage);
}

async function createEmptyBook() {
  const rootAccount = Account.create({
    guid: 'rootAccount',
    name: 'Root',
    type: 'ROOT',
  });
  await Account.upsert(rootAccount, ['guid']);

  await Book.upsert(
    {
      guid: 'maffinBook',
      fk_root: rootAccount.guid,
    },
    ['guid'],
  );
}

async function loadStockerPrices() {
  if (!isStaging()) {
    try {
      // We have to await because we use Price.upsert. Once we correct this,
      // saving Price already updates the local cache so we don't need await
      await insertTodayPrices();
    } catch (e) {
      console.warn(`Retrieving live prices failed ${e}`);
    }
  }
}
