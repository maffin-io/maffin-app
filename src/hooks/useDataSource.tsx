import React from 'react';
import { DataSource, QueryFailedError } from 'typeorm';
import initSqlJs from 'sql.js';
import { Settings } from 'luxon';
import pako from 'pako';

import { insertTodayPrices } from '@/lib/prices';
import useBookStorage from '@/hooks/useBookStorage';
import { migrate as migrateFromGnucash } from '@/lib/gnucash';
import {
  Account,
  Book,
  Commodity,
  Price,
  Split,
  Transaction,
} from '@/book/entities';
import { MIGRATIONS } from '@/book/migrations';
import { useQueryClient } from '@tanstack/react-query';
import type BookStorage from '@/lib/storage/BookStorage';
import { MaffinError, StorageError } from '@/helpers/errors';

export type DataSourceContextType = {
  datasource: DataSource | null,
  save: () => Promise<void>,
  importBook: (rawBook: Uint8Array) => Promise<void>,
  isLoaded: boolean,
};

const DATASOURCE: DataSource = new DataSource({
  type: 'sqljs',
  synchronize: true,
  logging: false,
  entities: [Account, Book, Commodity, Price, Split, Transaction],
  migrations: MIGRATIONS,
});

export const DataSourceContext = React.createContext<DataSourceContextType>({
  datasource: DATASOURCE,
  save: async () => {},
  importBook: async () => {},
  isLoaded: false,
});

export default function useDataSource(): DataSourceContextType {
  const { storage } = useBookStorage();
  const [isLoaded, setIsLoaded] = React.useState(DATASOURCE.isInitialized);
  const queryClient = useQueryClient();

  React.useEffect(() => {
    async function load() {
      let rawBook: Uint8Array;
      if (storage && !DATASOURCE.isInitialized && queryClient) {
        ([, rawBook] = await Promise.all([
          initSqlJs({ locateFile: file => `/${file}` }),
          storage.get(),
        ]));

        // Due to async nature, this value can change between this and the previous statement
        // as it's a shared global
        if (!DATASOURCE.isInitialized) {
          DATASOURCE.setOptions({
            type: 'sqljs',
            synchronize: true,
            logging: false,
            entities: [Account, Book, Commodity, Price, Split, Transaction],
            migrations: MIGRATIONS,
            extra: {
              queryClient,
            },
          });
          const start = performance.now();
          await DATASOURCE.initialize();

          if (rawBook.length) {
            await DATASOURCE.sqljsManager.loadDatabase(rawBook);
            await DATASOURCE.runMigrations();
          } else {
            await createEmptyBook();
          }
          const end = performance.now();
          console.log(`init datasource: ${end - start}ms`);
        }
        loadStockerPrices();
        setIsLoaded(true);
      }
    }

    load();
  }, [storage, queryClient]);

  if (!isLoaded) {
    return {
      datasource: DATASOURCE,
      save: async () => {},
      importBook: async () => {},
      isLoaded: false,
    };
  }

  return {
    datasource: DATASOURCE,
    save: () => save(storage as BookStorage),
    importBook: (rawData: Uint8Array) => importBook(storage as BookStorage, rawData),
    isLoaded,
  };
}

async function save(storage: BookStorage) {
  DATASOURCE.options.extra.queryClient.setQueryData(['state', 'isSaving'], true);
  await DATASOURCE.query('VACUUM');
  const rawBook = DATASOURCE.sqljsManager.exportDatabase();
  try {
    await storage.save(rawBook);
  } catch (e) {
    (e as MaffinError).show();
  } finally {
    DATASOURCE.options.extra.queryClient.setQueryData(['state', 'isSaving'], false);
  }
}

async function importBook(storage: BookStorage, rawData: Uint8Array) {
  if (process.env.NEXT_PUBLIC_ENV !== 'master') {
    Settings.now = () => Date.now();
  }
  let parsedData;
  try {
    parsedData = pako.ungzip(rawData);
  } catch (err) {
    parsedData = rawData;
  }

  const previousData = DATASOURCE.sqljsManager.exportDatabase();
  await DATASOURCE.sqljsManager.loadDatabase(parsedData);
  try {
    // Sadly loadDatabase doesn't check that the file is valid so we call this to trigger
    // errors if any
    await DATASOURCE.query('VACUUM');
  } catch (e) {
    let code = 'UNKNOWN';
    if (e instanceof QueryFailedError) {
      code = 'INVALID_FILE';
    }
    new StorageError((e as Error).message, code).show();
    await DATASOURCE.sqljsManager.loadDatabase(previousData);
    return;
  }

  // We probably should do this conditionally only for when it is a gnucash file
  await migrateFromGnucash();
  DATASOURCE.options.extra.queryClient.refetchQueries({
    queryKey: ['api'],
    type: 'all',
  });

  loadStockerPrices();
  await save(storage);
}

async function createEmptyBook() {
  const rootAccount = Account.create({
    guid: 'rootAccount',
    name: 'Root',
    type: 'ROOT',
  });
  await Account.upsert(rootAccount, ['guid']);

  await Book.upsert(
    {
      guid: 'maffinBook',
      fk_root: rootAccount.guid,
    },
    ['guid'],
  );
}

async function loadStockerPrices() {
  try {
    // We have to await because we use Price.upsert. Once we correct this,
    // saving Price already updates the local cache so we don't need await
    await insertTodayPrices();
  } catch (e) {
    console.warn(`Retrieving live prices failed ${e}`);
  }
}
