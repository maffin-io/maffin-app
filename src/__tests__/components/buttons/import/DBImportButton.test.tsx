import React from 'react';
import {
  render,
  screen,
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import ImportButton from '@/components/buttons/import/DBImportButton';
import { DataSourceContext } from '@/hooks';
import type { DataSourceContextType } from '@/hooks';

describe('DBImportButton', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('disables button when datasource not available', async () => {
    const { container } = render(
      <DataSourceContext.Provider value={{ isLoaded: false } as DataSourceContextType}>
        <ImportButton />
      </DataSourceContext.Provider>,
    );

    const e = await screen.findByRole('button');
    expect(e).toBeDisabled();
    expect(container).toMatchSnapshot();
  });

  it('renders as expected when datasource ready', async () => {
    render(
      <DataSourceContext.Provider value={{ isLoaded: true } as DataSourceContextType}>
        <ImportButton />
      </DataSourceContext.Provider>,
    );

    const e = await screen.findByRole('button');
    expect(e).not.toBeDisabled();
  });

  it('uploads and imports data into datasource', async () => {
    const mockOnImport = jest.fn();
    const mockImportBook = jest.fn();
    const file = new File(['hello'], 'hello.png', { type: 'image/png' });
    const user = userEvent.setup();

    render(
      <DataSourceContext.Provider
        value={{
          isLoaded: true,
          importBook: mockImportBook as Function,
        } as DataSourceContextType}
      >
        <ImportButton onImport={mockOnImport} />
      </DataSourceContext.Provider>,
    );

    const importButton = await screen.findByRole('button');
    await user.click(importButton);

    await userEvent.upload(screen.getByLabelText('importInput'), file);

    expect(mockImportBook).toBeCalledWith(new Uint8Array([104, 101, 108, 108, 111]));
    expect(mockOnImport).toBeCalled();
  });
});
