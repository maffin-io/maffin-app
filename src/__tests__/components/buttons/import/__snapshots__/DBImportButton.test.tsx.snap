// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DBImportButton disables button when datasource not available 1`] = `
<div>
  <button
    class=""
    disabled=""
    id="db-import-button"
    type="button"
  >
    <div
      class="card hover:text-cyan-600 dark:hover:text-white hover:shadow-xl"
    >
      <span
        class="float-right badge default -mt-2 -mr-2"
        data-tooltip-id="db-import-help"
      >
        ?
      </span>
      <div
        class="flex items-center"
      >
        <img
          alt=""
          class="m-0"
          data-nimg="1"
          decoding="async"
          height="65"
          loading="lazy"
          src="/_next/image?url=%2Fimg.jpg&w=256&q=75"
          srcset="/_next/image?url=%2Fimg.jpg&w=96&q=75 1x, /_next/image?url=%2Fimg.jpg&w=256&q=75 2x"
          style="color: transparent;"
          width="65"
        />
        <span
          class="inline-block align-middle"
        >
          Maffin DB
        </span>
      </div>
    </div>
  </button>
  <input
    aria-label="importInput"
    class="hidden"
    type="file"
  />
</div>
`;
